canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,864 @@
+                Aws::CloudWatch::Model::MetricDatum persistedAckLatencyDatum;
+                Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+                cwRequest.SetNamespace("KinesisVideoSDKCanary");
+
+                auto currentTimestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+                auto persistedAckLatency = (currentTimestamp - timeOfFragmentEndSent); // [milliseconds]
+                pushMetric("PersistedAckLatency", persistedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, persistedAckLatencyDatum, data->pDimensionPerStream, cwRequest);
+                LOG_DEBUG("Persisted Ack Latency: " << persistedAckLatency);
+                if (data->pCanaryConfig->useAggMetrics)
+                {
+                    pushMetric("PersistedAckLatency", persistedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, persistedAckLatencyDatum, data->pAggregatedDimension, cwRequest);
+
+                }
+                data->pCWclient->PutMetricDataAsync(cwRequest, onPutMetricDataResponseReceivedHandler);
+                break;
+            }
+            case FRAGMENT_ACK_TYPE_RECEIVED:
+            {
+                Aws::CloudWatch::Model::MetricDatum receivedAckLatencyDatum;
+                Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+                cwRequest.SetNamespace("KinesisVideoSDKCanary");
+
+                auto currentTimestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+                auto receivedAckLatency = (currentTimestamp - timeOfFragmentEndSent); // [milliseconds]
+                pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatencyDatum, data->pDimensionPerStream, cwRequest);
+                LOG_DEBUG("Received Ack Latency: " << receivedAckLatency);
+                if (data->pCanaryConfig->useAggMetrics)
+                {
+                    pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatencyDatum, data->pAggregatedDimension, cwRequest);
+                }
+                data->pCWclient->PutMetricDataAsync(cwRequest, onPutMetricDataResponseReceivedHandler);
+                break;
+            }
+            case FRAGMENT_ACK_TYPE_BUFFERING:
+            {
+                cout << "FRAGMENT_ACK_TYPE_BUFFERING callback invoked" << endl;
+                break;
+            }
+            case FRAGMENT_ACK_TYPE_ERROR:
+            {
+                cout << "FRAGMENT_ACK_TYPE_ERROR callback invoked" << endl;
+                break;
+            }
+        }
+    }
+}
+
+}  // namespace video
+}  // namespace kinesis
+}  // namespace amazonaws
+}  // namespace com;
+
+
+// add frame pts, frame index, original frame size, CRC to beginning of buffer
+VOID addCanaryMetadataToFrameData(PFrame pFrame)
+{
+    PBYTE pCurPtr = pFrame->frameData;
```

Comment:
Doing this would overwrite the gstreamer pFrame->frameData. You would have to memcopy the GStreamer frame data to the location starting from CANARY_METADATA_SIZE to ensure the header is appended at the start.